transport/socket_transport.go

```diff
@@ -0,0 +1,503 @@
+	defer s.acceptLoopExitGroup.Done()
+	for {
+		conn, err := s.listener.Accept()
+		if err != nil {
+			// Ok - was closed
+			break
+		}
+		c := &serverConnection{
+			s:    s,
+			conn: conn,
+		}
+		s.serverConnections.Store(c, struct{}{})
+		c.start()
+	}
+}
+
+func (s *SocketServer) removeConnection(conn *serverConnection) {
+	s.serverConnections.Delete(conn)
+}
+
+type serverConnection struct {
+	s          *SocketServer
+	conn       net.Conn
+	closeGroup sync.WaitGroup
+	lock       sync.Mutex
+	closed     bool
+}
+
+func (c *serverConnection) start() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.closeGroup.Add(1)
+	common.Go(c.readLoop)
+}
+
+func (c *serverConnection) readLoop() {
+	defer c.readPanicHandler()
+	defer c.closeGroup.Done()
+	if err := readMessage(c.conn, c.handleMessage); err != nil {
+		// Closed connection errors are normal on server shutdown - we ignore them
+		ignoreErr := false
+		if err == io.EOF {
+			ignoreErr = true
+		} else if ne, ok := err.(net.Error); ok {
+			msg := ne.Error()
+			ignoreErr = strings.Contains(msg, "use of closed network connection")
+		}
+		if !ignoreErr {
+			log.Errorf("error in reading from server connection: %v", err)
+		}
+		if err := c.conn.Close(); err != nil {
+			// Ignore
+		}
+	}
+	c.cleanUp()
+}
+
+func (c *serverConnection) cleanUp() {
+	c.s.removeConnection(c)
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.closed = true
+}
+
+func (c *serverConnection) readPanicHandler() {
+	// We use a custom panic handler as we don't want the server to panic and crash if it receives a malformed
+	// request which has insufficient bytes in the buffer which would cause a runtime error: index out of range panic
+	if r := recover(); r != nil {
+		log.Errorf("failure in connection readLoop: %v", r)
+		if err := c.conn.Close(); err != nil {
+			// Ignore
+		}
+		c.cleanUp()
+	}
+}
+
+func readMessage(conn net.Conn, messageHandler func([]byte) error) error {
+	buff := make([]byte, readBuffSize)
+	var err error
+	var readPos, n int
+	for {
+		// read the message size
+		bytesRequired := 4 - readPos
+		if bytesRequired > 0 {
+			n, err = io.ReadAtLeast(conn, buff[readPos:], bytesRequired)
+			if err != nil {
+				break
+			}
+			readPos += n
+		}
+		totSize := 4 + int(binary.BigEndian.Uint32(buff))
+		bytesRequired = totSize - readPos
+		if bytesRequired > 0 {
+			// If we haven't already read enough bytes, read the entire message body
+			if totSize > len(buff) {
+				// buffer isn't big enough, resize it
+				nb := make([]byte, totSize)
+				copy(nb, buff)
+				buff = nb
+			}
+			n, err = io.ReadAtLeast(conn, buff[readPos:], bytesRequired)
+			if err != nil {
+				break
+			}
+			readPos += n
+		}
+		// Note that the buffer is reused so it's up to the protocol structs to copy any data in the message such
+		// as records, uuid, []byte before the call to handleMessage returns
+		err = messageHandler(buff[4:totSize])
+		if err != nil {
+			break
+		}
+		remainingBytes := readPos - totSize
```

Comment:
On the first iteration this will be negative no?

Reply:
Since read pos is unitialized

Reply:
readPos will be initialised by this point. Note, this is not new code, it's moved from the Kafka server which is used for every Kafka request, so I'd be surprised if there was such an error in it.